### ISSUE TITLE
Can't parse nested @styles

### ISSUE THREAD

**Issue by jamietre** [jsdom/jsdom#1374]
I realize this is not actually a real CSS rule, but this is a convention used by PDFReactor (a 3rd party tool we use) and browsers don't have a problem with it, but jsdom throws an error when trying to parse this.

```
@page {
    @top-center {
        content: element(headerIdentifier);
    }
} 

Error: Unexpected } (line 5, char 1)
Error: Could not load link: "http://localhost:5000/report-server/styles/styles2.css"
    at ..\report-server\node_modules\jsdom\lib\jsdom\browser\resource-loader.js:36:21
```

Basically, it seems like nested at-rules can't be parsed. I suspect there's nothing technically "wrong" with the way jsdom/cssstyle is handling this, since there are not actually valid nested @rules, but I can't think of an advantage to failing to parse a style sheet which browsers are able to parse.

**Comment by jamietre** [comment]
I take that back, there do seem to be valid cases for nested at rules, e.g.


@font-feature-values Font One { /* How to activate nice-style in Font One */
  @styleset {
    nice-style: 12;
  }
}

**Comment by domenic** [comment]
Is this just informational output on the virtual console, or does this actually crash your program/call a callback with an error? If it's the former, that is expected; features we don't support (or more accurately, [CSSOM](https://github.com/NV/CSSOM/) does not support) trigger an error. If it's the latter, this is a bug.

**Comment by jamietre** [comment]
It doesn't crash the program, but it fails to load the style sheet in its entirety.

**Comment by domenic** [comment]
Yeah... I guess this needs to be fixed in the CSSOM module, so it's better about ignoring things it doesn't understand. I have merge permissions over there if you can manage to work up a PR...

**Comment by jamietre** [comment]
Never looked at it before... but how complicated could it be?

**Comment by jamietre** [comment]
.. so the parser in CSSOM is a little rough, it's a pretty deliberately coded scanner. It doesn't support nested structures at all and is definitely not spec-driven. It probably wouldn't be too hard to fix the problem I'm having right now but I am guessing there are many other edge cases..

At the same time, I just tried out the other relatively mature ones I could find, and they either broke on the same constructs or on other things (`reworkcss/css`. `yahoo/css-js`, 'cwdog/cssparser')

In fact the only one I found that was capable of parsing my entire style sheet, and generally, of handling nested structures was this, which looks pretty new, but also looks very well written and well documented:

https://github.com/PolymerLabs/shady-css-parser

Any thought to using a different css parser? Or what about decoupling the parser itself it from JSDOM and just specify a tree builder API so it wouldn't be hard for end users to change implementations if they need to? Not sure how much of a priority this is for the project but just some thoughts...

**Comment by Sebmaster** [comment]
> Any thought to using a different css parser?

If it supports more stuff and is tested / documented more - sure.

> This parser is not spec-compliant, however it will parse all spec-compliant CSS.

This may be a problem, but since it improves the overall situation right now, it's fine.

> Malformed CSS will be parsed by this parser as closely as possible to the way a browser would parse it.

makes me think it's a good thing anyways.

> Or what about decoupling the parser itself it from JSDOM and just specify a tree builder API so it wouldn't be hard for end users to change implementations if they need to? Not sure how much of a priority this is for the project but just some thoughts...

We're actually going in a somewhat opposite direction in our parsing area - actually reducing the amount of interoperability and requiring other packages to emulate on of the supported parsers, since (at least with parsing) we require a lot of fine-tuning to (exactly) match the spec(s).

**Comment by domenic** [comment]
The biggest problem with switching out cssom would be [replacing the actual CSSOM interfaces it implements](https://github.com/tmpvar/jsdom/blob/6f3b7b64c3726f63e4dd62bfa4120b6f6807d026/lib/jsdom/level2/style.js#L24-L30). You'd presumably have to implement those from scratch and then work to send the parser output into new instances of them.

**Comment by jamietre** [comment]
I was thinking just leave CSSOM as the object model implementation but not use its parser. `shady-css-parser` doesn't actually have an object model, it just returns an AST as json.

**Comment by domenic** [comment]
I'm not sure how you could do that---i.e., I can't imagine that CSSOM takes as its input an AST---but if it's possible, that would be cool.

**Comment by jamietre** [comment]
.. btw when I said cssom doesn't support nested structions, this is really just the parser, the object model seems to have considered them. 

As far as how it could be done, I was thinking you wouldn't pass css parsing directly to CSSOM, but do it with some abstraction that uses shady-css (or whatever) to parse, and returns a structure made with CSSOM objects.

**Comment by jamietre** [comment]
.. an alternative, too, would be alter CSSOM so it's not tightly coupled with its own parser and allow passing it an AST directly in some proscribed format.

@sebastian I certainly understand the motivation to keep control over as much as possible to maintain quality. I don't think decoupling the parser itself violates this conceptually, though, simply because it's a very well defined subsystem.  If there's a de-facto implementation available (or one eventually becomes available) then I would think you'd prefer to use it than maintain your own. 

But I definitely agree that the object model itself should be kept in-house.

For now the quick and dirty solution to my problem seems to be to use shady-css to generate an AST, filter out the offending CSS, and pass it back into jsdom :)

If you guys generally agree to allowing interop with a different parser to build the object model, I would be interested in writing the code.

**Comment by Sebmaster** [comment]
> I certainly understand the motivation to keep control over as much as possible to maintain quality.

It wasn't really about the quality, but parsing requires so much interoperation with element creation, that abstracting over different modules for parsing made it impossible to actually behave according to spec. If that's not a problem with styles - better for us.

> If there's a de-facto implementation available (or one eventually becomes available) then I would think you'd prefer to use it than maintain your own.

Oh I'd love for us to have to maintain less code. Sadly, npm modules don't really want to provide a proper object model most of the time :stuck_out_tongue: 

> If you guys generally agree to allowing interop with a different parser to build the object model, I would be interested in writing the code.

Sure. We'll eventually either have to get our own CSS OM or fix up cssom to properly follow IDL specs anyways I assume (they are probably not totally correct).

**Comment by jamietre** [comment]
> It wasn't really about the quality, but parsing requires so much interoperation with element creation, that abstracting over different modules for parsing made it impossible to actually behave according to spec. If that's not a problem with styles - better for us.

The inefficient way (of just parsing an already-build AST into your own CSS OM) wouldn't be hard.  The more efficient way of building it during parsing looks also possible since `shady-css-parser` seems to expose its own AST building blocks.

Anyway I'll give it a shot over the next few days and see how it works out.

**Comment by jamietre** [comment]
It shis a problem, btw? Not an MIT license

http://polymer.github.io/LICENSE.txt

**Comment by domenic** [comment]
The license is not a problem.

Looking forward to seeing what you can work out!!
